pkg/cluster/removeprivatednszone.go

```diff
@@ -27,125 +28,85 @@ func (m *manager) removePrivateDNSZone(ctx context.Context) error {
+	clusterVersionIsLessThan4_4, err := version.ClusterVersionIsLessThan4_4(ctx, m.configcli)
 	if err != nil {
 		m.log.Print(err)
 		return nil
 	}
 
-	if len(nodes.Items) != machineCount {
-		m.log.Printf("cluster has %d nodes but %d under MCPs, not removing private DNS zone", len(nodes.Items), machineCount)
+	if clusterVersionIsLessThan4_4 {
+		m.log.Printf("cluster version < 4.4, not removing private DNS zone")
 		return nil
 	}
 
-	cv, err := m.configcli.ConfigV1().ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
-	if err != nil {
-		return err
-	}
-	v, err := version.GetClusterVersion(cv)
-	if err != nil {
+	if err = utilnet.UpdateDNSs(ctx, m.configcli.ConfigV1().DNSes(), resourceGroupID); err != nil {
 		m.log.Print(err)
 		return nil
 	}
 
-	if v.Lt(version.NewVersion(4, 4)) {
-		// 4.3 uses SRV records for etcd
-		m.log.Printf("cluster version < 4.4, not removing private DNS zone")
+	if err := utilnet.RemoveZones(ctx, m.virtualNetworkLinks, m.privateZones, zones, resourceGroup); err != nil {
+		m.log.Print(err)
 		return nil
 	}
+	return nil
+}
 
-	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		dns, err := m.configcli.ConfigV1().DNSes().Get(ctx, "cluster", metav1.GetOptions{})
-		if err != nil {
-			return err
-		}
-
-		if dns.Spec.PrivateZone == nil ||
-			!strings.HasPrefix(strings.ToLower(dns.Spec.PrivateZone.ID), strings.ToLower(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID)) {
-			return nil
-		}
+// sameNumberOfNodesAndMachines returns true if the cluster has the same number of nodes and total machines,
+// and an error if any. Otherwise it returns false and nil.
+func (m *manager) sameNumberOfNodesAndMachines(ctx context.Context) (bool, error) {
+	machineConfigPoolList, err := m.mcocli.MachineconfigurationV1().MachineConfigPools().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return false, err
+	}
 
-		dns.Spec.PrivateZone = nil
+	totalMachines, err := totalMachinesInTheMCPs(machineConfigPoolList.Items)
+	if err != nil {
+		return false, err
+	}
 
-		_, err = m.configcli.ConfigV1().DNSes().Update(ctx, dns, metav1.UpdateOptions{})
-		return err
-	})
+	nodes, err := m.kubernetescli.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
 	if err != nil {
-		m.log.Print(err)
-		return nil
+		return false, err
 	}
 
-	for _, zone := range zones {
-		err = m.deletePrivateDNSVirtualNetworkLinks(ctx, *zone.ID)
-		if err != nil {
-			m.log.Print(err)
-			return nil
-		}
+	nNodes := len(nodes.Items)
+	if nNodes != totalMachines {
+		return false, fmt.Errorf("cluster has %d nodes but %d under MCPs, not removing private DNS zone", nNodes, totalMachines)
+	}
 
-		r, err := azure.ParseResourceID(*zone.ID)
-		if err != nil {
-			m.log.Print(err)
-			return nil
+	return true, nil
+}
+
+// totalMachinesInTheMCPs returns the total number of machines in the machineConfigPools
+// and an error, if any.
+func totalMachinesInTheMCPs(machineConfigPools []mcv1.MachineConfigPool) (int, error) {
```

Comment:
I'd like to see unit tests for new functions, especially ones with clearly defined scope.

Reply:
Done 😄